### ISSUE TITLE
Noise scheduler issue.

### ISSUE THREAD

**Issue by reverse-sky** [LingxiaoShawn/USD3#1]
Hello, I am a user who has been benefiting greatly from your code. I wanted to reach out because it seems that there might be an issue with the noise scheduler's beta_t (I’ve confirmed that alphabar_t is working correctly). I’ve made some modifications by borrowing the format of a standard diffusion model. I’m sharing this in case it might be helpful.

Below is the code snippet for the changes I made.

```python
from functools import partial
import numpy as np

# --------------------------------------------------------------------------------
def linear_beta_schedule(timesteps):
    scale = 1000 / timesteps
    beta_start = scale * 0.0001
    beta_end = scale * 0.02
    return np.linspace(0 , 1, 100, dtype=np.float32)

def cosine_beta_schedule(timesteps, s = 0.008):
    """
    cosine schedule
    as proposed in https://openreview.net/forum?id=-NEXDKk8gZ
    """
    steps = timesteps + 1
    x = np.linspace(0, steps, steps)
    alphas_cumprod = np.cos(((x / steps) + s) / (1 + s) * np.pi * 0.5) ** 2
    alphas_cumprod = alphas_cumprod / alphas_cumprod[0]
    betas = 1 - (alphas_cumprod[1:] / alphas_cumprod[:-1])
    return np.clip(betas, a_min = 0, a_max = 0.999)

##################



# https://github.com/lucidrains/denoising-diffusion-pytorch/blob/5989f4c77eafcdc6be0fb4739f0f277a6dd7f7d8/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L308
def noise_schedule(schedule_type:str= "cosine",
                   N:int = 1000, # N=0 means continuous 
                   sqrt = True,
                   Tmax:float =1,
                   **kwargs):
    """
    N : diffusion Step
    N >= t_step >=0
    """
    # for the previous step, we need to add 1 in diffusion step
    if schedule_type == "cosine":      
        betas = cosine_beta_schedule(N+1)
    elif schedule_type == "linear":
        betas = linear_beta_schedule(N+1)
    elif schedule_type == "exponential":
        raise NotImplementedError
    elif schedule_type == "constant":
        raise NotImplementedError
    else:
        raise NotImplementedError
    to_torch = partial(torch.tensor, dtype=torch.float32)
    
    alphas = 1. - betas
    alphas_cumprod = np.cumprod(alphas, axis=0)
    betas = to_torch(betas)
    alphas_cumprod = to_torch(alphas_cumprod)    
    # optional
    if sqrt:
        alphas_cumprod = torch.sqrt(alphas_cumprod)  


    return alphas_cumprod, betas
    
    ...
    class UnifiedDiscreteDiffusion:
    """
    Unified discrete-time and continuous-time discrete diffusion model.
    The code is closely following the paper's notation and implementation.
    Understanding the paper's equations is necessary to follow the code.
    We work with categorical labels directly, instead of one-hot representation. 

    - Notice that the 'conditional_mask' argument is used for both conditioning and padding part. 
    """
    def __init__(self, 
                 num_steps, 
                 num_classes, 
                 noise_schedule_type, 
                 noise_schedule_args,
                 simplified_max_val=1
                 noise_sqrt = False
                 ):
        self.num_steps = num_steps                      # 0 indicates using continuous-time diffusion 
        self.num_classes = num_classes
        self.noise_schedule_type = noise_schedule_type
        self.noise_schedule_args = noise_schedule_args  # args passed to noise_schedule
        self.simplified_max_val = simplified_max_val

       ### add this part only to save the noise scheduler
        self.alphas_cumprod, self.betas  = noise_schedule(self.noise_schedule_type, 
                                                      self.num_steps, 
                                                      sqrt = noise_sqrt,
                                                      Tmax = 1)
  ......
  
    @torch.no_grad()
    def get_alphabar_beta(self, t, s=False): 
        # t : (B,)
        alphabar_t = self.alphas_cumprod.to(t.device).index_select(0, t.long())
        beta_t     = self.betas.to(t.device).index_select(0, t.long())
        return alphabar_t, beta_t
    
    
```

**Comment by reverse-sky** [comment]
![image](https://github.com/user-attachments/assets/071bcbbe-025e-49ce-87ab-d76a6758a026)

**# Edited $\beta_t$**

This figure demonstrates the issue with the original $\beta_t$ function. The first plot shows the $\beta_t$ before modification, where it is evident that the values deviate significantly from the expected range of [0,1] for a diffusion model's $\beta_t$. Although only the cosine function is plotted here, similar issues are present with other $\beta_t$ functions as well. After applying the modified function, $\beta_t$ values fall within the correct range, and you can see them plotted alongside the original $\bar{\alpha}_{t}$.

**Figure [3]** shows a comparison between the original $\bar{\alpha}_{t}$ and the modified $\bar{\alpha}_{t}$, specifically when the square root operation is included.

**# sqrt Existence**

In the DDPM paper, the $\alpha$ values are used with the square root operation. However, in the D3PM paper, the square root was not used when defining transitions. Therefore, the default setting is set to False, but I have made it optional for use. Previously, the $\bar{\alpha}_{t}$ values with the square root were plotted, as shown in **Figure [4]**.

**Comment by LingxiaoShawn** [comment]
Please notice that our beta is not the same as the DDPM paper. Checkout section 3.2 of our paper to find the definition of beta.
